repose-aggregator/components/filters/keystone-v2/src/main/scala/org/openrepose/filters/keystonev2/KeystoneV2Filter.scala

```diff
@@ -0,0 +1,360 @@
+              Reject(SC_INTERNAL_SERVER_ERROR, failure = Some(x))
+          }
+        } getOrElse {
+          //Don't have an auth token to validate
+          Reject(SC_FORBIDDEN, Some("Auth token not found in headers"))
+        }
+      }
+
+      //Handle the result of the filter
+      result match {
+        case rejection: Reject =>
+          val message: Option[String] = rejection match {
+            case Reject(_, Some(x), _) => Some(x)
+            case Reject(code, None, Some(failure)) => Some(failure.getMessage)
+            case _ => None
+          }
+          //Handle delegation if necessary
+          Option(config.getDelegating) match {
+            case Some(delegating) =>
+              logger.debug(s"Delegating with status ${rejection.status}")
+              val delegationHeaders = buildDelegationHeaders(rejection.status,
+                "keystone-v2",
+                message.getOrElse("Failure in the Keystone v2 filter").replace("\n", ""),
+                delegating.getQuality)
+
+              delegationHeaders foreach { case (key, values) =>
+                values foreach { value =>
+                  request.addHeader(key, value)
+                }
+              }
+              request.addHeader(OpenStackServiceHeader.IDENTITY_STATUS.toString, IdentityStatus.Indeterminate.toString)
+
+              chain.doFilter(request, response)
+            case None =>
+              logger.debug(s"Rejecting with status ${rejection.status}")
+              response.setHeader(CommonHttpHeader.WWW_AUTHENTICATE.toString, s"Keystone uri=${config.getIdentityService.getUri}")
+              message match {
+                case Some(m) =>
+                  logger.debug(s"Rejection message: $m")
+                  response.sendError(rejection.status, m)
+                case None => response.sendError(rejection.status)
+              }
+          }
+
+        case p: Pass =>
+          //If the token validation passed and we're configured to do more things, we can do additional authorizations
+          //Modify the request to add stuff
+          p.headersToAdd.foreach { case (k, v) =>
+            request.addHeader(k, v)
+          }
+          chain.doFilter(request, response)
+
+          logger.trace("Keystone v2 filter processing response...")
+          val wwwAuthenticateValue = response.getHeader(CommonHttpHeader.WWW_AUTHENTICATE)
+
+          response.getStatus match {
+            case HttpServletResponse.SC_UNAUTHORIZED | HttpServletResponse.SC_FORBIDDEN =>
+              Option(config.getDelegating) foreach { delegating =>
+                  // If in the case that the origin service supports delegated authentication
+                  // we should then communicate to the client how to authenticate with us
+                  response.addHeader(CommonHttpHeader.WWW_AUTHENTICATE, s"Keystone uri=${config.getIdentityService.getUri}")
+              }
+            case HttpServletResponse.SC_NOT_IMPLEMENTED =>
```

Comment:
I guess it is. I don't actually know the specifics of the old contract, but it seems to me like we would get this if we forwarded an indeterminate identity status and the origin service responded with a not implemented implying that it can't handle authentication.